rclcpp/include/rclcpp/intra_process_manager_impl.hpp

```diff
@@ -184,6 +186,7 @@ class IntraProcessManagerState : public IntraProcessManagerStateBase
     size_t & size
   )
   {
+    std::lock_guard<std::mutex> lock(runtime_mutex_);
```

Comment:
A mutex per item sounds like overkill; I would leave it as is.

I hadn't dug into the data types being handled, and didn't realize that you still need exclusion after pulling the items out of `publishers_`.